File: app/clone/structure/Collection.scala
package clone.structure

import com.google.inject.Inject
import com.mongodb.client.model.IndexOptions
import org.bson.types.ObjectId
import play.api.libs.json.{JsObject, Json}
import services.{ConfigurationService, MongoService, SectionId}

import scala.util.Try

class Collection (conf: ConfigurationService, val db: String, val coll: String, primary: MongoService, secondary: MongoService) {
  var collStats: CollectionStats = primary.getCollectionStats(db, coll)
  var previousId: Option[String] = None

  /**
    * In charge of preparing the collection to synchronize, returns the various CollectionParts needed to synchronize everything
    */
  def prepareSync(): List[CollectionPart] = {

    // Drop / create the destination collection
    checkCollection()

    // Add indexes
    copyIndexes()

    // Get various section ids
    val sectionIds = listSectionIds
    if(sectionIds.size < 2) {
      throw new Exception("We should always have at least 2 seeds")
    }

    // Create and return the list of CollectionParts for the current Collection
    var previousSectionId = sectionIds.head
    sectionIds.drop(1).map(sectionId => {
      val collPart = if(db == "local" && coll == "oplog.rs") new OplogCollectionPart(conf, this, primary, secondary, previousSectionId, sectionId, sectionIds.size)
      else new BasicCollectionPart(conf, this, primary, secondary, previousSectionId, sectionId, sectionIds.size)

      previousSectionId = sectionId
      collPart
    })
  }

  /**
    * Load the seeds we want to use for the synchronisation. Return a list of tuples, each tuple represent a range,
        the first value is the start of it, the second value, the end of it.
    * @return
    */
  def listSectionIds: List[SectionId] = {
    val firstSectionId = SectionId(Option(new ObjectId("0"*24)))
    val lastSectionId = SectionId(Option(new ObjectId("f"*24)))

    // First, we need to be sure that we have an _id and if that's an objectid, otherwise we cannot use the same technique.
    // The oplog should only be tailed by one thread at a time, so we want to be sure to never create seeds for it.
    if(Try{primary.getObjectId(primary.first(db, coll).get).get}.isFailure || (db == "local" && coll == "oplog.rs")) {
      return List(SectionId(None),SectionId(None))
    }

    // Number of seeds we would like
    val quantity = conf.maximumSeeds
    if(collStats.count <= 100 * quantity) { // Arbitrarily, we decide it's useless to use a lot of seeds if we only have a small number of documents
      return List(firstSectionId, lastSectionId)
    }

    val sectionIds = primary.sectionIds(db, coll, quantity).sortBy(_.oid.get.getTimestamp)

    // TODO: In the future, if we want to be smart and allow retry of a failed sync, we should take the previous seeds
    // stored in the mongosync database, then make a simple query to see up to where they went

    // Always add the first and last seed
    List(firstSectionId) ::: sectionIds ::: List(lastSectionId)
  }

  /**
    * Specific checks before writing to a collection
    */
  def checkCollection(): Unit = {
    // Stats about the destination collection
    var destinationStats: Option[CollectionStats] = if (secondary.listDatabases.contains(db) && secondary.listCollections(db).contains(coll)) {
      Option(secondary.getCollectionStats(db, coll))
    } else {
      None
    }

    // For the internal databases, we want to remove them by default, to avoid any problem (one exception: the oplog)
    if(destinationStats.isDefined && db == "local" && coll != "oplog.rs") { // Note: this was deactivated in python
      secondary.dropCollection(db, coll)
      destinationStats = None
    }

    if(collStats.capped && destinationStats.isEmpty) {
      var capped_max_size = collStats.maxSize
      val capped_max = collStats.max

      if(db == "local" && coll == "oplog.rs") {
        //Special case, we do not necessarily want to keep the same oplog size as the other node
        capped_max_size = conf.mongoOplogSize * 1024 * 1024 * 1024
      }

      secondary.createCollection(db, coll, true, capped_max, capped_max_size)
    }
  }

  /**
    * It is better to copy indexes directly, before copying the data. That way we directly have the TTL working, but we also
        do not need to read all data at the end to create the indexes (if Mongo is on a NFS mount, you will lose a lot of
        time just for that).
        Even if the insert performance will be "worst", at the end, it should not matter a lot with the multi-threading copy
        of mongosync
    */
  def copyIndexes(): Unit = {
    primary.getCollectionIndexes(db, coll).foreach(index => {
      val key = (index \ "key").as[JsObject]
      val options = new IndexOptions()
      // TODO: Handle the index options
      secondary.createCollectionIndex(db, coll, key, options)
    })
  }
}

File: src/core/clone/Collection.py
from src.core.service.Mongo import Mongo
import time
import pymongo
from bson.objectid import ObjectId

class Collection:
    def __init__(self, configuration, db, coll):
        self.configuration = configuration
        self.db = db
        self.coll = coll
        self.mongo_primary = Mongo(configuration, is_primary=True)
        self.mongo_secondary = Mongo(configuration, is_primary=False)

        self.coll_stats = self.mongo_primary.collection_stats(db=self.db, coll=self.coll)
        self.previous_id = None

    """
        In charge of preparing the collection to synchronize, returns the various seeds we should use. 
    """
    def prepare_sync(self):
        average_object_size = self.coll_stats['avgObjSize']
        expected_documents  = self.coll_stats['count'] # It can increase but it's not a problem, it's only used for logging

        # Drop / Create the destination collection
        self.check_collection()

        # Get the various seeds
        seeds = self.list_seeds()
        if len(seeds) == 0:
            print("We should always have at least 2 seeds.")
            raise ValueError("Invalid seed number. Failure.")

        # Create and return the list of inputs necessary to create CollectionPart
        collection_parts = []
        previous_seed = seeds[0]
        for seed in seeds[1:]:
            collection_parts.append({
                'db': self.db,
                'coll': self.coll,
                'seed_start': previous_seed,
                'seed_end': seed,
                'total_seeds': len(seeds)
            })
            previous_seed = seed
        return collection_parts



    """
        Load the seeds we want to use for the synchronisation. Return a list of tuples, each tuple represent a range,
        the first value is the start of it, the second value, the end of it.
    """
    def list_seeds(self):
        # First, we need to be sure that we have an _id and if that's an objectid, otherwise we cannot use the same technique.
        # The oplog should only be tailed by one thread at a time, so we want to be sure to never create seeds for it.
        id_type = self.mongo_primary.id_type(self.db, self.coll)
        if id_type['has_id'] is False or id_type['is_object_id'] is False or (self.db == "local" and self.coll == "oplog.rs"):
            return [None, None]


        # Number of seeds we would like
        quantity = self.configuration.internal_maximum_seeds()
        if self.coll_stats['count'] <= 100*quantity: # Arbitrarily, we decide it's useless to use a lot of seeds if we only have a small number of documents
            return [({'_id':ObjectId('0'*24)},{'_id':ObjectId('f'*24)})]

        # Get various seeds
        seeds = self.mongo_primary.section_ids(self.db, self.coll, quantity=quantity)

        # TODO: In the future, if we want to be smart and allow retry of a failed sync, we should take the previous seeds
        # stored in the mongosync database, then make a simple query to see up to where they went

        # We order them to be able to return ranges. the ObjectId already allows to compare values.
        seeds = sorted(seeds, key=lambda seed: seed['_id'])

        # Always add the first and last seed
        seeds = [{'_id':ObjectId('0'*24)}] + seeds
        seeds.append({'_id':ObjectId('f'*24)})

        return seeds

    """
        Specific checks before writing to a collection
    """
    def check_collection(self):
        # Stats about the optional collection
        if self.db in self.mongo_secondary.list_databases() and self.coll in self.mongo_secondary.list_collections(self.db):
            destination_stats = self.mongo_secondary.collection_stats(db=self.db, coll=self.coll)
        else:
            destination_stats = {}

        # For internal db, we want to remove them by default, to avoid any problem (one exception: the oplog)
        if len(destination_stats) != 1 and self.db == 'local':
            if self.coll != 'oplog.rs' and False:
                # Not possible to drop every db
                self.mongo_secondary.drop(self.db, self.coll)
                self.destination_stats = {}

        # Optionally create a capped collection, but we only do that if it didn't exist before
        if self.coll_stats['capped'] is True and len(destination_stats) == 0:
            capped_max_size = self.coll_stats.get('maxSize', -1)
            capped_max = self.coll_stats.get('max', -1)
            if self.coll_stats['ns'] == 'local.oplog.rs':
                # Special case, we do not necessarily want to keep the same oplog size as the other node
                capped_max_size = self.configuration.mongo_oplog_size() * (1024 ** 3)

            if capped_max_size == -1:
                capped_max_size = None
            if capped_max == -1:
                capped_max = None

            self.mongo_secondary.create_collection(self.db, self.coll, capped=True, max=capped_max, max_size=capped_max_size)


    def __str__(self):
        return 'Collection:'+self.db+'.'+self.coll

    def __repr__(self):
        return self.__str__()
